{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "1a813897",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_fasta_file(input_file):\n",
    "    \"\"\"This parses FASTA files and compiles them into a dictionary \n",
    "    with the identity tag as the keys\n",
    "    and the nucleotide or amino acid sequences as the values\n",
    "    \"\"\"\n",
    "\n",
    "    parsed_seqs = {}\n",
    "\n",
    "    f = open(input_file)\n",
    "\n",
    "    curr_seq_id = None\n",
    "    curr_seq = []\n",
    "\n",
    "    for line in f:\n",
    "\n",
    "            line = line.strip()\n",
    "\n",
    "            if line.startswith(\">\"):\n",
    "                if curr_seq_id is not None:\n",
    "                    parsed_seqs[curr_seq_id] = ''.join(curr_seq)\n",
    "\n",
    "                curr_seq_id = line[1:]\n",
    "                curr_seq = []\n",
    "                continue\n",
    "\n",
    "            curr_seq.append(line)\n",
    "\n",
    "    parsed_seqs[curr_seq_id] = ''.join(curr_seq)\n",
    "    \n",
    "    print('Parse Complete')\n",
    "\n",
    "    reference_protein = list(parsed_seqs)[0]\n",
    "    reference_sequence = list(parsed_seqs.values())[0]\n",
    "    \n",
    "    return parsed_seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ef556dec",
   "metadata": {},
   "outputs": [],
   "source": [
    "def analyze_nucleotide_abundance(parsed_seqs, species):\n",
    "    \"\"\"computes the abundance of every mononucleotide and dinucleotide in a genome\n",
    "    input_file -- a FASTA file to parse, placing every gene has been placed in a dictionary\n",
    "    with the id tag as the key and the residue sequence as the value\n",
    "    \"\"\"\n",
    "\n",
    "    genome_string = \"\"\n",
    "    #creates an empty string labeled \"genome_string\"\n",
    "    \n",
    "    sequences = parsed_seqs\n",
    "            \n",
    "    species_column = []\n",
    "    nuc_id = []\n",
    "    nuc_count = []\n",
    "    nuc_percent = []\n",
    "    \n",
    "    for identity_tag, sequence in sequences.items():\n",
    "        \n",
    "        genome_string += sequence\n",
    "        # compiles all residues from an annotated genome into a single string labeled genome_string\n",
    "    \n",
    "    genome_string = genome_string.upper()\n",
    "    # converts all nucleotide letters in genome_string to upper case\n",
    "    \n",
    "    total_count = len(genome_string)\n",
    "    #print(\"Total Length: \",total_count)\n",
    "\n",
    "    a_count = genome_string.count(\"A\")\n",
    "    a_percent = round(a_count / total_count, 4) * 100\n",
    "    species_column.append(species)\n",
    "    nuc_id.append('A')\n",
    "    nuc_count.append(a_count)\n",
    "    nuc_percent.append(a_percent)\n",
    "    #print(f\"Number of A's: \",a_count,\",\",a_percent,\"%\")\n",
    "    \n",
    "    t_count = genome_string.count(\"T\")\n",
    "    t_percent = round(t_count / total_count, 4) * 100\n",
    "    species_column.append(species)\n",
    "    nuc_id.append('T')\n",
    "    nuc_count.append(t_count)\n",
    "    nuc_percent.append(t_percent)\n",
    "    #print(f\"Number of T's: \",t_count,\",\",t_percent,\"%\")\n",
    "    \n",
    "    c_count = genome_string.count(\"C\")\n",
    "    c_percent = round(c_count / total_count, 4) * 100\n",
    "    species_column.append(species)\n",
    "    nuc_id.append('C')\n",
    "    nuc_count.append(c_count)\n",
    "    nuc_percent.append(c_percent)\n",
    "    #print(f\"Number of C's: \",c_count,\",\",c_percent,\"%\")\n",
    "    \n",
    "    g_count = genome_string.count(\"G\")\n",
    "    g_percent = round(g_count / total_count, 4) * 100\n",
    "    species_column.append(species)\n",
    "    nuc_id.append('G')\n",
    "    nuc_count.append(g_count)\n",
    "    nuc_percent.append(g_percent)\n",
    "    #print(f\"Number of G's: \",g_count,\",\",g_percent,\"%\")\n",
    "    \n",
    "    aa_count = genome_string.count(\"AA\")\n",
    "    aa_percent = round(aa_count / total_count, 4) * 100\n",
    "    species_column.append(species)\n",
    "    nuc_id.append('AA')\n",
    "    nuc_count.append(aa_count)\n",
    "    nuc_percent.append(aa_percent)\n",
    "    #print(f\"Number of AA's: \",aa_count,\",\",aa_percent,\"%\")\n",
    "        \n",
    "    at_count = genome_string.count(\"AT\")\n",
    "    at_percent = round(at_count / total_count, 4) * 100\n",
    "    species_column.append(species)\n",
    "    nuc_id.append('AT')\n",
    "    nuc_count.append(at_count)\n",
    "    nuc_percent.append(at_percent)\n",
    "    #print(f\"Number of AT's: \",at_count,\",\",at_percent,\"%\")\n",
    "        \n",
    "    ac_count = genome_string.count(\"AC\")\n",
    "    ac_percent = round(ac_count / total_count, 4) * 100\n",
    "    species_column.append(species)\n",
    "    nuc_id.append('AC')\n",
    "    nuc_count.append(ac_count)\n",
    "    nuc_percent.append(ac_percent)\n",
    "    #print(f\"Number of AC's: \",ac_count,\",\",ac_percent,\"%\")\n",
    "        \n",
    "    ag_count = genome_string.count(\"AG\")\n",
    "    ag_percent = round(ag_count / total_count, 4) * 100\n",
    "    species_column.append(species)\n",
    "    nuc_id.append('AG')\n",
    "    nuc_count.append(ag_count)\n",
    "    nuc_percent.append(ag_percent)\n",
    "    #print(f\"Number of AG's: \",ag_count,\",\",ag_percent,\"%\")\n",
    "        \n",
    "    ta_count = genome_string.count(\"TA\")\n",
    "    ta_percent = round(ta_count / total_count, 4) * 100\n",
    "    species_column.append(species)\n",
    "    nuc_id.append('TA')\n",
    "    nuc_count.append(ta_count)\n",
    "    nuc_percent.append(ta_percent)\n",
    "    #print(f\"Number of TA's: \",ta_count,\",\",ta_percent,\"%\")\n",
    "        \n",
    "    tt_count = genome_string.count(\"TT\")\n",
    "    tt_percent = round(tt_count / total_count, 4) * 100\n",
    "    species_column.append(species)\n",
    "    nuc_id.append('TT')\n",
    "    nuc_count.append(tt_count)\n",
    "    nuc_percent.append(tt_percent)\n",
    "    #print(f\"Number of TT's: \",tt_count,\",\",tt_percent,\"%\")\n",
    "        \n",
    "    tc_count = genome_string.count(\"TC\")\n",
    "    tc_percent = round(tc_count / total_count, 4) * 100\n",
    "    species_column.append(species)\n",
    "    nuc_id.append('TC')\n",
    "    nuc_count.append(tc_count)\n",
    "    nuc_percent.append(tc_percent)\n",
    "    #print(f\"Number of TC's: \",tc_count,\",\",tc_percent,\"%\")\n",
    "        \n",
    "    tg_count = genome_string.count(\"TG\")\n",
    "    tg_percent = round(tg_count / total_count, 4) * 100\n",
    "    species_column.append(species)\n",
    "    nuc_id.append('TG')\n",
    "    nuc_count.append(tg_count)\n",
    "    nuc_percent.append(tg_percent)\n",
    "    #print(f\"Number of TG's: \",tg_count,\",\",tg_percent,\"%\")\n",
    "            \n",
    "    ca_count = genome_string.count(\"CA\")\n",
    "    ca_percent = round(ca_count / total_count, 4) * 100\n",
    "    species_column.append(species)\n",
    "    nuc_id.append('CA')\n",
    "    nuc_count.append(ca_count)\n",
    "    nuc_percent.append(ca_percent)\n",
    "    #print(f\"Number of CA's: \",ca_count,\",\",ca_percent,\"%\")\n",
    "        \n",
    "    ct_count = genome_string.count(\"CT\")\n",
    "    ct_percent = round(ct_count / total_count, 4) * 100\n",
    "    species_column.append(species)\n",
    "    nuc_id.append('CT')\n",
    "    nuc_count.append(ct_count)\n",
    "    nuc_percent.append(ct_percent)\n",
    "    #print(f\"Number of CT's: \",ct_count,\",\",ct_percent,\"%\")\n",
    "        \n",
    "    cc_count = genome_string.count(\"CC\")\n",
    "    cc_percent = round(cc_count / total_count, 4) * 100\n",
    "    species_column.append(species)\n",
    "    nuc_id.append('CC')\n",
    "    nuc_count.append(cc_count)\n",
    "    nuc_percent.append(cc_percent)\n",
    "    #print(f\"Number of CC's: \",cc_count,\",\",cc_percent,\"%\")\n",
    "        \n",
    "    cg_count = genome_string.count(\"CG\")\n",
    "    cg_percent = round(cg_count / total_count, 4) * 100\n",
    "    species_column.append(species)\n",
    "    nuc_id.append('CG')\n",
    "    nuc_count.append(cg_count)\n",
    "    nuc_percent.append(cg_percent)\n",
    "    #print(f\"Number of CG's: \",cg_count,\",\",cg_percent,\"%\")\n",
    "                \n",
    "    ga_count = genome_string.count(\"GA\")\n",
    "    ga_percent = round(ga_count / total_count, 4) * 100\n",
    "    species_column.append(species)\n",
    "    nuc_id.append('GA')\n",
    "    nuc_count.append(ga_count)\n",
    "    nuc_percent.append(ga_percent)\n",
    "    #print(f\"Number of GA's: \",ga_count,\",\",ga_percent,\"%\")\n",
    "        \n",
    "    gt_count = genome_string.count(\"GT\")\n",
    "    gt_percent = round(gt_count / total_count, 4) * 100\n",
    "    species_column.append(species)\n",
    "    nuc_id.append('GT')\n",
    "    nuc_count.append(gt_count)\n",
    "    nuc_percent.append(gt_percent)\n",
    "    #print(f\"Number of GT's: \",gt_count,\",\",gt_percent,\"%\")\n",
    "    \n",
    "    gc_count = genome_string.count(\"GC\")\n",
    "    gc_percent = round(gc_count / total_count, 4) * 100\n",
    "    species_column.append(species)\n",
    "    nuc_id.append('GC')\n",
    "    nuc_count.append(gc_count)\n",
    "    nuc_percent.append(gc_percent)\n",
    "    #print(f\"Number of GC's: \",gc_count,\",\",gc_percent,\"%\")\n",
    "            \n",
    "    gg_count = genome_string.count(\"GG\")\n",
    "    gg_percent = round(gg_count / total_count, 4) * 100\n",
    "    species_column.append(species)\n",
    "    nuc_id.append('GG')\n",
    "    nuc_count.append(gg_count)\n",
    "    nuc_percent.append(gg_percent)\n",
    "    #print(f\"Number of GG's: \",gg_count,\",\",gg_percent,\"%\")\n",
    "    \n",
    "    # prints the number and proportion of mononucleotides and dinucleotides\n",
    "    return species_column, nuc_id, nuc_count, nuc_percent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1e93a07e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def analyze_dinucleotide_ratio(parsed_seqs, species):\n",
    "    \"\"\"Computes the abundance of CG dinucleotides and GC content in each gene of a genome and the ratio between them\n",
    "    then places these values into lists, along with the id tag of each gene and the species it comes from\n",
    "    parsed_seqs -- a parsed FASTA file in which every gene has been placed in a dictionary\n",
    "    with the id tag as the key and the residue sequence as the value\n",
    "    species -- binomial name of the species being analyzed in string form, e.g. 'Thelohanellus_kitauei'\n",
    "    \"\"\"\n",
    "    \n",
    "    sequences = parsed_seqs\n",
    "    \n",
    "    species_column = []\n",
    "    gene_ids = []\n",
    "    cg_counts = []\n",
    "    gc_counts = []\n",
    "    gc_contents = []\n",
    "    dinuc_ratios = []\n",
    "    \n",
    "    # creates empty lists for the five dataframe columns\n",
    "    \n",
    "    for identity_tag, sequence in sequences.items():\n",
    "        \n",
    "        sequence = sequence.upper()\n",
    "        \n",
    "        species_column.append(species)\n",
    "        gene_ids.append(identity_tag)\n",
    "        \n",
    "        # places species name and gene id tags in lists\n",
    "        \n",
    "        cg_count = sequence.count(\"CG\")\n",
    "        cg_counts.append(cg_count)\n",
    "        \n",
    "        gc_count = sequence.count(\"GC\")\n",
    "        gc_counts.append(gc_count)\n",
    "        \n",
    "        gc_content = sequence.count(\"G\") + sequence.count(\"C\")\n",
    "        gc_contents.append(gc_content)\n",
    "        \n",
    "        # calculates the CG and GC dinucleotide counts and GC content for each gene and places them in lists\n",
    "        \n",
    "        if gc_count == 0:\n",
    "            cg_to_gc_ratio = 0\n",
    "        else:\n",
    "            cg_to_gc_ratio = round(cg_count / gc_count, 4)\n",
    "            \n",
    "        dinuc_ratios.append(cg_to_gc_ratio)\n",
    "        \n",
    "        # calculates the ratio of CGs to GCs and places them in a list\n",
    "        # while preventing division by zero errors\n",
    "    \n",
    "    return species_column, gene_ids, cg_counts, gc_counts, gc_contents, dinuc_ratios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "8bd08f3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_gene_datatable(input_file, species, directory):\n",
    "    \"\"\"creates a pandas datatable out of the lists returned by the analyze_dinucleotide_ratio function\n",
    "    then saves this datatable as a csv file to a specific directory\n",
    "    input_file -- a FASTA file to parse, placing every gene has been placed in a dictionary\n",
    "    with the id tag as the key and the residue sequence as the value\n",
    "    species -- binomial name of the species being analyzed in string form, e.g. 'Thelohanellus_kitauei'\n",
    "    directory -- a directory in string form to which the csv data table will be saved, e.g. 'C:Downloads'\n",
    "    \"\"\"\n",
    "    \n",
    "    import pandas as pd\n",
    "    import os.path\n",
    "    \n",
    "    parsed_seqs = parse_fasta_file(input_file)\n",
    "    \n",
    "    # retrieves a dictionary of parsed sequences from the parse_fasta_file function\n",
    "    \n",
    "    species_column, gene_ids, cg_counts, gc_counts, gc_contents, dinuc_ratios = analyze_dinucleotide_ratio(parsed_seqs, species)\n",
    "    \n",
    "    # retrieves lists of dinucleotide abundances and ratios from the analyze_dinucleotide_ratio function\n",
    "    \n",
    "    data = {'Species': species_column,\n",
    "            'Gene ID': gene_ids,\n",
    "            'CG Dinucleotide Counts': cg_counts,\n",
    "            'GC Dinucleotide Counts': gc_counts,\n",
    "            'GC Content': gc_contents,\n",
    "            'CG to GC Ratio': dinuc_ratios}\n",
    "    \n",
    "    df = pd.DataFrame(data)\n",
    "    \n",
    "    # creates a dataframe using each list as a column with data\n",
    "    \n",
    "    df.to_csv(os.path.join(f'{directory}',f'{species}_dinucleotide_analysis.csv'))\n",
    "    \n",
    "    # writes csv file to specified directory\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "abfd8505",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_genome_datatable(input_file, species, directory):\n",
    "    \"\"\"creates a pandas datatable out of the lists returned by the analyze_dinucleotide_abundance function\n",
    "    then saves this datatable as a csv file to a specific directory\n",
    "    input_file -- a FASTA file to parse, placing every gene has been placed in a dictionary\n",
    "    with the id tag as the key and the residue sequence as the value\n",
    "    species -- binomial name of the species being analyzed in string form, e.g. 'Thelohanellus_kitauei'\n",
    "    directory -- a directory in string form to which the csv data table will be saved, e.g. 'C:Downloads'\n",
    "    \"\"\"\n",
    "    \n",
    "    import pandas as pd\n",
    "    import os.path\n",
    "    \n",
    "    parsed_seqs = parse_fasta_file(input_file)\n",
    "    \n",
    "    # retrieves a dictionary of parsed sequences from the parse_fasta_file function\n",
    "    \n",
    "    species_column, nuc_id, nuc_count, nuc_percent = analyze_nucleotide_abundance(parsed_seqs, species)\n",
    "    \n",
    "    # retrieves lists of nucleotide proportions from the analyze_nucleotide_abundance function\n",
    "    \n",
    "    data = {'Species': species_column,\n",
    "            'Nucleotides': nuc_id,\n",
    "            'Nucleotide Counts': nuc_count,\n",
    "            'Nucleotide Percentages': nuc_percent}\n",
    "    \n",
    "    df = pd.DataFrame(data)\n",
    "    \n",
    "    # creates a dataframe using each list as a column with data\n",
    "    \n",
    "    df.to_csv(os.path.join(f'{directory}',f'{species}_nucleotide_proportions.csv'))\n",
    "    \n",
    "    # writes csv file to specified directory\n",
    "    \n",
    "    return df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
